### ISSUE TITLE
Nvmf initiator: nvme_rdma.c:1678:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x25feda0,

### ISSUE THREAD

**Issue by fshenx** [spdk/spdk#641]
I press Ctrl+c on the nvmf target when I running bdevperf test case on the initiator, and then, the initiator will print error messages.

## Expected Behavior
The normal exit
## Current Behavior
initiator:
```
# ./test/bdev/bdevperf/bdevperf -c bdev.conf -q 128 -o 524288 -w randrw -M 50  -t 30
I/O size of 524288 is greater than zero copy threshold (65536).
Zero copy mechanism will not be used.
Starting SPDK v19.01-pre / DPDK 18.11.0 initialization...
[ DPDK EAL parameters: bdevperf --no-shconf -c 0x1 --base-virtaddr=0x200000000000 --file-prefix=spdk_pid1152449 ]
EAL: Detected 44 lcore(s)
EAL: Detected 2 NUMA nodes
EAL: No free hugepages reported in hugepages-1048576kB
EAL: Probing VFIO support...
app.c: 625:spdk_app_start: *NOTICE*: Total cores available: 1
reactor.c: 298:_spdk_reactor_run: *NOTICE*: Reactor started on core 0
Running I/O for 30 seconds...
nvme_rdma.c:1678:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x25feda0, Response Index 42045664 (12): transport retry counter exceeded
nvme_rdma.c:1678:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x25feda0, Response Index 12 (5): Work Request Flushed Error
nvme_rdma.c:1678:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x25feda0, Response Index 42047008 (5): Work Request Flushed Error
nvme_rdma.c:1678:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x25feda0, Response Index 42047200 (5): Work Request Flushed Error
nvme_rdma.c:1678:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x25feda0, Response Index 42047392 (5): Work Request Flushed Error
nvme_rdma.c:1678:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x25feda0, Response Index 42047584 (5): Work Request Flushed Error
nvme_rdma.c:1678:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x25feda0, Response Index 42047776 (5): Work Request Flushed Error
nvme_rdma.c:1678:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x25feda0, Response Index 42047968 (5): Work Request Flushed Error
nvme_rdma.c:1678:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x25feda0, Response Index 42048160 (5): Work Request Flushed Error
nvme_rdma.c:1678:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x25feda0, Response Index 42048352 (5): Work Request Flushed Error
nvme_rdma.c:1678:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x25feda0, Response Index 42048544 (5): Work Request Flushed Error
nvme_rdma.c:1678:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x25feda0, Response Index 42048736 (5): Work Request Flushed Error
nvme_rdma.c:1678:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x25feda0, Response Index 42048928 (5): Work Request Flushed Error
nvme_rdma.c:1678:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x25feda0, Response Index 42049120 (5): Work Request Flushed Error
nvme_rdma.c:1678:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x25feda0, Response Index 42049312 (5): Work Request Flushed Error
nvme_rdma.c:1678:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x25feda0, Response Index 42049504 (5): Work Request Flushed Error
nvme_rdma.c:1678:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x25feda0, Response Index 42049696 (5): Work Request Flushed Error
nvme_rdma.c:1678:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x25feda0, Response Index 42049888 (5): Work Request Flushed Error
nvme_rdma.c:1678:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x25feda0, Response Index 42050080 (5): Work Request Flushed Error
nvme_rdma.c:1678:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x25feda0, Response Index 42050272 (5): Work Request Flushed Error
nvme_rdma.c:1678:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x25feda0, Response Index 42050464 (5): Work Request Flushed Error
nvme_rdma.c:1678:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x25feda0, Response Index 42050656 (5): Work Request Flushed Error
nvme_rdma.c:1678:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x25feda0, Response Index 42050848 (5): Work Request Flushed Error
nvme_rdma.c:1678:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x25feda0, Response Index 42051040 (5): Work Request Flushed Error
nvme_rdma.c:1678:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x25feda0, Response Index 42051232 (5): Work Request Flushed Error
nvme_rdma.c:1678:nvme_rdma_qpair_process_completions: *ERROR*: CQ error on Queue Pair 0x25feda0, Response Index 42051424 (5): Work Request Flushed Error
```

target:
```
# ./app/nvmf_tgt/nvmf_tgt -c nvmf.conf.in
Starting SPDK v19.01-pre / DPDK 18.11.0 initialization...
[ DPDK EAL parameters: nvmf --no-shconf -c 0x1 --base-virtaddr=0x200000000000 --file-prefix=spdk_pid5060 ]
EAL: Detected 44 lcore(s)
EAL: Detected 2 NUMA nodes
EAL: No free hugepages reported in hugepages-1048576kB
EAL: Probing VFIO support...
app.c: 625:spdk_app_start: *NOTICE*: Total cores available: 1
app.c: 551:spdk_app_setup_trace: *NOTICE*: Tracepoint Group Mask 0x01 specified.
app.c: 555:spdk_app_setup_trace: *NOTICE*: Use 'spdk_trace -s nvmf -p 5060' to capture a snapshot of events at runtime.
app.c: 557:spdk_app_setup_trace: *NOTICE*: Or copy /dev/shm/nvmf_trace.pid5060 for offline analysis/debug.
reactor.c: 298:_spdk_reactor_run: *NOTICE*: Reactor started on core 0
^Cthread.c: 249:spdk_thread_exit: *ERROR*: thread reactor_0 still has channel for io_device rdma_transport
```

## Steps to Reproduce
1. target:   ./app/nvmf_tgt/nvmf_tgt -c nvmf.conf.in
nvmf.conf.in:
```
[Global]
 TpointGroupMask 0x01

[Bdev]

[Malloc]
  NumberOfLuns 8
  LunSizeInMB 64

[AIO]

[Pmem]

[Nvmf]
  AcceptorPollRate 10000

[Transport]
  Type RDMA
  MaxQueueDepth 2048

[Nvme]
  TransportID "trtype:PCIe traddr:0000:06:00.0" Nvme0
  RetryCount 4
  TimeoutUsec 0
  ActionOnTimeout None
  AdminPollRate 100000
  HotplugEnable No

[Subsystem1]
  NQN nqn.2016-06.io.spdk:cnode1
  Listen RDMA 192.168.1.3:4420
  AllowAnyHost Yes
  Host nqn.2016-06.io.spdk:init
  SN SPDK00000000000001
  MaxNamespaces 20
  Namespace Nvme0n1 1
```
2. initiator:  ./test/bdev/bdevperf/bdevperf -c bdev.conf -q 128 -o 524288 -w randrw -M 50  -t 30
bdev.conf:
```
[Nvme]
        TransportID "trtype:RDMA adrfam:IPv4 traddr:192.168.1.3 trsvcid:4420 subnqn:nqn.2016-06.io.spdk:cnode1" Nvme0
```
3. target:  Ctrl+c
4. initiator: after a while, it will print error messages

##Environment
target:
commit id: ce75af214
OS: Ubuntu 18.04
kernel:4.16.7

initiator:
commit id: ce75af214
OS: Fedora release 27
kernel: 4.16.7

**Comment by Seth5141** [comment]
@fshenx  I don't think this is actually an issue. When we ctrl-C the trrget, we initialize the shutdown behavior of the whole system, including all qpairs. This means that we have to abort all outstanding I/O from the target and move the qpair into error state. Then on the initiator side, they get failed completions for all of the I/O that was outstanding. The error message that the initiator prints out is simply to indicate those I/O that failed when we shut down the target. From the initiator side, getting an error completion is an error, but it's the best the target can do when shutting down.
Does the initiator hang after printing those error messages, or is there some other problem that prevents  bdevperf from exiting and freeing the memory associated with it? I think that those things would be issues, but if the only side effect here is that the initiator prints error completions when the target disconnects unexpectedly, I would say that both the initiator and target are doing their jobs. The target is properly notifying all incoming connections and aborting in-flight I/O before a sudden disconnect, and the initiator is reporting those failed I/O.

**Comment by optimistyzy** [comment]
@fshenx I used the latest master branch 55be9a57a619e275826314e90e8c7c1d6348abc7 to test, and see that there is no issue related with the target, but bdevperf does not exit. I tested in the same host.

**Comment by optimistyzy** [comment]
I think that the issue is in bdev layer, we do not have the timeout policy. When the connection is broken, bdev_nvme_poll is still polling(it is registered a poller), and we know those I/Os will never be completed, but it is not aborted.

**Comment by shenfurong2015** [comment]
@optimistyzy 
 yes, there is no issue on the target, but bdevperf does not exit.

**Comment by fshenx** [comment]
@optimistyzy
yes, there is no issue on the target, but bdevperf does not exit.

**Comment by optimistyzy** [comment]
I think that are some issues in bdev_nvme module. When the connection is broken (e.g., the NVMe-oF target is shutting down). So the situation is that:  The users submit lot of bdev I/Os, and the call back function provided by the users are never executed. Currently I did not have a solution for this issue yet. But I have some ideas to conduct:

1  Need to update the code in bdev_nvme_poll. If the return value of spdk_nvme_qpair_process_completions is -1, we need not call this again and again, and we need to mark all the I/Os submitted on this I/O channel failed.
2  We should have some policys in bdev layer to timeout the bdev io (may be in the general layer).  It means that the bdev io is submitted without return for a time, we need to fail those I/Os and call the user provided call back.

Currently in our code, we do not test this exceptional case very good both for local NVMe devices and with remote NVMe-of target. So especially in bdev_nvme module, we did not consider such exceptional case. Definitely, there is no issue with the NVMe-of target. But the problem is in our bdev NVMe module. We did not handle the qpair failure(e.g., due to connection broken) well.
